evap/staff/importers/json.py

```diff
@@ -0,0 +1,417 @@
+
+        log += self._make_stats("New Courses", self.new_courses)
+        log += self._make_stats("New Evaluations", self.new_evaluations)
+        log += self._make_stats("Updated Courses", self.updated_courses)
+        log += self._make_stats("Updated Evaluations", self.updated_evaluations)
+        log += self._make_stats("Attempted Changes", self.attempted_changes)
+
+        log += self._make_heading("Warnings")
+        log += self._make_total(len(self.warnings))
+        for warning in self.warnings:
+            log += f"- {warning.obj}: {warning.message}\n"
+
+        return log
+
+    def send_mail(self):
+        subject = "[EvaP] JSON importer log"
+
+        managers = UserProfile.objects.filter(groups__name="Manager", email__isnull=False)
+        if not managers:
+            return
+        mail = EmailMultiAlternatives(
+            subject,
+            self.get_log(),
+            settings.SERVER_EMAIL,
+            [manager.email for manager in managers],
+        )
+        mail.send()
+
+
+class JSONImporter:
+    DATETIME_FORMAT = "%d.%m.%Y %H:%M:%S"
+
+    def __init__(self, semester: Semester, default_course_end: str) -> None:
+        self.semester = semester
+        self.default_course_end = default_course_end
+        self.user_profile_map: dict[str, UserProfile] = {}
+        self.course_type_cache: dict[str, CourseType] = {
+            import_name.strip().lower(): course_type
+            for course_type in CourseType.objects.all()
+            for import_name in course_type.import_names
+        }
+        self.program_cache: dict[str, Program] = {
+            import_name.strip().lower(): program
+            for program in Program.objects.all()
+            for import_name in program.import_names
+        }
+        self.course_map: dict[str, Course] = {}
+        self.statistics = ImportStatistics()
+
+    def _choose_responsibles(self, user_profiles: list[UserProfile]) -> list[UserProfile]:
+        user_profiles_by_len = {}
+        for user_profile in user_profiles:
+            user_profile_len = len(user_profile.title)
+            user_profiles_by_len.setdefault(user_profile_len, []).append(user_profile)
+        for key in sorted(user_profiles_by_len.keys(), reverse=True):
+            return user_profiles_by_len[key]
+        return []
+
+    def _filter_user_profiles(self, user_profiles: list[UserProfile]) -> list[UserProfile]:
+        return list(filter(lambda p: p.email not in settings.NON_RESPONSIBLE_USERS, user_profiles))
+
+    def _get_course_type(self, name: str) -> CourseType:
+        lookup = name.strip().lower()
+        if lookup in self.course_type_cache:
+            return self.course_type_cache[lookup]
+
+        course_type, __ = CourseType.objects.get_or_create(name_de=name, defaults={"name_en": name})
+        self.course_type_cache[name] = course_type
+        return course_type
```

Comment:
This is always fully filled on `__init__`, so we should never not have an entry, right? (Same below for the non-covered line in `_get_program`)

```suggestion
        lookup = name.strip().lower()
        return self.course_type_cache[lookup]
```